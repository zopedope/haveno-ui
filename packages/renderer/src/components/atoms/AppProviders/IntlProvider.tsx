// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { useEffect, useMemo, useState } from "react";
import { IntlProvider as ReacIntlProvider } from "react-intl";
import type { FC } from "react";
import { LangPack, SupportedLocales } from "@src/constants/lang";
import type { LangKeys } from "@src/constants/lang";

const DEFAULT_LOCALE = SupportedLocales.EN;

export const IntlProvider: FC = ({ children }) => {
  const [messages, setMessages] = useState<{ [key in LangKeys]: string }>();
  const locale = useMemo(
    () =>
      navigator.language in SupportedLocales
        ? navigator.language
        : DEFAULT_LOCALE,
    [navigator.language]
  );

  useEffect(() => {
    setMessages(LangPack[locale] ?? LangPack[DEFAULT_LOCALE]);
  }, [locale]);

  return (
    <ReacIntlProvider locale={locale} messages={messages}>
      {children}
    </ReacIntlProvider>
  );
};
