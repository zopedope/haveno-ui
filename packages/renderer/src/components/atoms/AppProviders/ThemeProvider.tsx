// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { MantineProvider, Global } from "@mantine/core";
import type { FC } from "react";
import { themeOverride, globalStyles } from "@src/theme";

export const ThemeProvider: FC = ({ children }) => {
  return (
    <MantineProvider theme={themeOverride}>
      <Global styles={globalStyles} />
      {children}
    </MantineProvider>
  );
};
