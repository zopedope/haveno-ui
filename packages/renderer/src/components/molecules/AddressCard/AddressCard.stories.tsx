// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { Stack } from "@mantine/core";
import type { ComponentStory, ComponentMeta } from "@storybook/react";
import { AddressCard } from "./AddressCard";

export default {
  title: "molecules/AddressCard",
  component: AddressCard,
} as ComponentMeta<typeof AddressCard>;

const Template: ComponentStory<typeof AddressCard> = (args) => {
  return (
    <Stack>
      <AddressCard
        {...args}
        primary={true}
        qrModalProps={{ target: "#root" }}
      />
      <AddressCard
        {...args}
        primary={false}
        qrModalProps={{ target: "#root" }}
      />
    </Stack>
  );
};

export const Default = Template.bind({});

Default.args = {
  label: "Primary Address",
  primary: true,
  address:
    "44tgg3TkQ2jGRDabB5cjbNWDF7PKDBKqw2bsjgRRCQSThiE15ePWk6kJFH7YWnPKR88JQB8WwDX34TwfYnhWVeT1J1rC6b7",
};
