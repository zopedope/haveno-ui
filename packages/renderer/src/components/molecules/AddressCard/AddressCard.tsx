// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { FormattedMessage, useIntl } from "react-intl";
import QRCode from "react-qr-code";
import { showNotification } from "@mantine/notifications";
import { useModals } from "@mantine/modals";
import { useClipboard } from "@mantine/hooks";
import { Box, createStyles, Group, SimpleGrid, Skeleton } from "@mantine/core";
import type { OpenConfirmModal } from "@mantine/modals/lib/context";
import { DetailItem } from "@atoms/DetailItem";
import { Button } from "@atoms/Buttons";
import { Anchor } from "@atoms/Typography";
import { DetailItemCard } from "@atoms/DetailItemCard";
import { LangKeys } from "@constants/lang";
import { useSetDownloadQRCode } from "@hooks/haveno/useSetDownloadQRCode";
import { Modals } from "@constants/modals";

interface AddressCardProps {
  label?: string;
  address: string;
  primary?: boolean;
  qrModalProps?: OpenConfirmModal;
}

const COPY_TEXT_TIMEOUT = 500;

export function AddressCard({
  label,
  address,
  primary = false,
  qrModalProps,
}: AddressCardProps) {
  const modals = useModals();
  const { classes } = useStyles();
  const { formatMessage } = useIntl();

  const { mutateAsync: downloadQRCode } = useSetDownloadQRCode();

  const clipboard = useClipboard({ timeout: COPY_TEXT_TIMEOUT });

  const handleCopyClick = () => {
    clipboard.copy(address);
  };
  const handleQRDownloadClick = (addressCode: string) => {
    downloadQRCode(addressCode).then(() => {
      showNotification({
        color: "green",
        message: formatMessage({
          id: LangKeys.AddressCardQRCodeSavedSuccessNotif,
          defaultMessage: "The QR code has been saved successfully.",
        }),
      });
      modals.closeModal(Modals.QRCodeAddress);
    });
  };
  const handleQRClick = () => {
    const modalId = modals.openModal({
      id: Modals.QRCodeAddress,
      children: (
        <AddressCardQRModalContent
          address={address}
          onQRDownloadClick={handleQRDownloadClick}
          onReturnClick={() => modals.closeModal(modalId)}
        />
      ),
      labels: { confirm: "Confirm", cancel: "Cancel" },
      size: "lg",
      withCloseButton: false,
      ...qrModalProps,
    });
  };

  return (
    <DetailItemCard label={label} primary={primary}>
      <Group noWrap className={classes.contentGroup}>
        <Box className={classes.address}>{address}</Box>

        <Group noWrap className={classes.addressBtns}>
          <Anchor onClick={handleCopyClick} underline>
            {!clipboard.copied ? (
              <FormattedMessage
                id={LangKeys.AddressCardCopyBtn}
                defaultMessage="Copy"
              />
            ) : (
              <FormattedMessage
                id={LangKeys.AddressCardCopiedBtn}
                defaultMessage="Copied"
              />
            )}
          </Anchor>

          <Anchor onClick={handleQRClick} underline>
            <FormattedMessage
              id={LangKeys.AddressCardQRBtn}
              defaultMessage="QR"
            />
          </Anchor>
        </Group>
      </Group>
    </DetailItemCard>
  );
}

type AddressCardSkeletonProps = Pick<AddressCardProps, "label" | "primary">;

export function AddressCardSkeleton({
  label,
  primary,
}: AddressCardSkeletonProps) {
  const { classes } = useStyles();

  return (
    <DetailItemCard label={label} primary={primary}>
      <Box className={classes.address}>
        <Skeleton
          width="80%"
          height={8}
          mt="xs"
          sx={(theme) => ({
            "&:before": {
              backgroundColor: primary
                ? theme.colors.gray[1]
                : theme.colors.gray[0],
            },
            "&:after": {
              backgroundColor: primary
                ? theme.colors.gray[4]
                : theme.colors.gray[3],
            },
          })}
        />
      </Box>
    </DetailItemCard>
  );
}

interface AddressCardQRModalContentProps {
  address: string;
  onQRDownloadClick?: (address: string) => void;
  onReturnClick?: () => void;
}

function AddressCardQRModalContent({
  address,
  onQRDownloadClick,
  onReturnClick,
}: AddressCardQRModalContentProps) {
  const { classes } = useStyles();
  const { formatMessage } = useIntl();

  const handleQRDownloadClick = () => {
    onQRDownloadClick && onQRDownloadClick(address);
  };
  return (
    <Box className={classes.qrModalRoot}>
      <DetailItem
        classNames={{
          content: classes.qrModalAddress,
          label: classes.qrModalAddressLabel,
        }}
        label={formatMessage({
          id: LangKeys.MyWalletQRModalPrimaryAddress,
          defaultMessage: "Primary Address",
        })}
      >
        {address}
      </DetailItem>

      <Box className={classes.qrRoot}>
        <QRCode value={address} size={370} />
      </Box>

      <SimpleGrid cols={2} mt="xl">
        <Button flavor="neutral" onClick={onReturnClick}>
          <FormattedMessage
            id={LangKeys.MyWalletQRModalReturnBtn}
            defaultMessage="Return"
          />
        </Button>

        <Button onClick={handleQRDownloadClick}>
          <FormattedMessage
            id={LangKeys.MyWalletQRModalDownloadQRBtn}
            defaultMessage="Download QR"
          />
        </Button>
      </SimpleGrid>
    </Box>
  );
}

const useStyles = createStyles((theme) => ({
  contentGroup: {
    minWidth: "100%",
  },
  address: {
    textOverflow: "ellipsis",
    overflow: "hidden",
    whiteSpace: "nowrap",
    width: "100%",
  },
  addressBtns: {
    marginLeft: "auto",
  },
  qrModalRoot: {
    padding: "1.5rem",
  },
  qrRoot: {
    marginTop: theme.spacing.xl * 1.5,
    marginBottom: theme.spacing.xl * 1.5,
    textAlign: "center",
  },
  qrModalAddress: {
    fontSize: theme.fontSizes.lg,
  },
  qrModalAddressLabel: {
    fontSize: theme.fontSizes.xs,
  },
}));
