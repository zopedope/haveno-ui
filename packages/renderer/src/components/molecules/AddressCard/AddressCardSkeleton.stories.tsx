// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { Stack } from "@mantine/core";
import type { ComponentStory, ComponentMeta } from "@storybook/react";
import { AddressCardSkeleton } from "./AddressCard";

export default {
  title: "molecules/AddressCardSkeleton",
  component: AddressCardSkeleton,
} as ComponentMeta<typeof AddressCardSkeleton>;

const Template: ComponentStory<typeof AddressCardSkeleton> = (args) => {
  return (
    <Stack>
      <AddressCardSkeleton {...args} primary={false} />
      <AddressCardSkeleton {...args} primary={true} />
    </Stack>
  );
};

export const Default = Template.bind({});

Default.args = {
  label: "Primary Address",
  primary: true,
};
