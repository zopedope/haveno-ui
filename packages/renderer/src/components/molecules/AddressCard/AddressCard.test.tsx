// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import { AddressCard } from "./AddressCard";
import { AppProviders } from "@atoms/AppProviders";

describe("molecules::AddressCard", () => {
  it("renders without exploding", () => {
    const { asFragment, unmount } = render(
      <AppProviders>
        <AddressCard address="Address here" label="Address label" />
      </AppProviders>
    );
    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  it("renderes the address hash code.", () => {
    const { unmount } = render(
      <AppProviders>
        <AddressCard address="ASDASDQWEKOOQLMWEM" />
      </AppProviders>
    );
    expect(screen.queryByText("ASDASDQWEKOOQLMWEM")).toMatchSnapshot();
    unmount();
  });

  it("renderes the address label.", () => {
    const { unmount } = render(
      <AppProviders>
        <AddressCard label="Primary Address" address="ASDASDQWEKOOQLMWEM" />
      </AppProviders>
    );
    expect(screen.queryByText("Primary Address")).toMatchSnapshot();
    unmount();
  });
});
