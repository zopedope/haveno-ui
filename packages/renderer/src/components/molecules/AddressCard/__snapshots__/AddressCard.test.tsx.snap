// Vitest Snapshot v1

exports[`molecules::AddressCard > renderes the address hash code. 1`] = `
<div
  class="mantine-Group-child mantine-2w2ufd"
>
  ASDASDQWEKOOQLMWEM
</div>
`;

exports[`molecules::AddressCard > renderes the address label. 1`] = `
<div
  class="mantine-Text-root mantine-DetailItem-label mantine-1ece26k"
>
  Primary Address
</div>
`;

exports[`molecules::AddressCard > renders without exploding 1`] = `
<DocumentFragment>
  <div
    class="mantine-Group-root mantine-chiw4t"
  >
    <div
      class="mantine-Stack-root mantine-DetailItem-root mantine-Group-child mantine-1a4fzbd"
    >
      <div
        class="mantine-Text-root mantine-DetailItem-label mantine-1ece26k"
      >
        Address label
      </div>
      <div
        class="mantine-Text-root mantine-DetailItem-content mantine-1tq817k"
      >
        <div
          class="mantine-Group-root mantine-1i932k4"
        >
          <div
            class="mantine-Group-child mantine-2w2ufd"
          >
            Address here
          </div>
          <div
            class="mantine-Group-root mantine-Group-child mantine-ze53qg"
          >
            <a
              class="mantine-Text-root mantine-Anchor-root mantine-Group-child mantine-odvvvx"
            >
              Copy
            </a>
            <a
              class="mantine-Text-root mantine-Anchor-root mantine-Group-child mantine-odvvvx"
            >
              QR
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
